FUSETOOLS-3030 - adding ability to create new REST operations

bfitzpat opened the pull request:
# Pull Request Checklist

After this checklist is all checked or PR provides explanations for possible pass-through, please put the label "Ready for review" on the PR.


## General

- [x] Did you use the Jira Issue number in the commit comments?
- [x] Did you set meaningful commit comments on each commit?
- [x] Did you sign off all commits for this PR? (git commit -s -m "jira issue number - your commit comment")

## Functional

- [ ] Did the CI job report a successful build?
- [ ] Is the non-happy path working, too?

## Maintainability

- [x] Are all Sonar reported issues fixed or can they be ignored?
- [x] Is there no duplicated code?
- [x] Are method-/class-/variable-names meaningful?

## Tests

- [ ] Are there unit-tests?
- [ ] Are there integration tests (or at least a jira to tackle these)?
- [ ] Do we need a new UI test?

## Legal

- [x] Have you used the correct file header copyright comment?
- [x] Have you used the correct year in the headers of new files?

bfitzpat commented:
This one will require the other PRs to be merged before it - especially https://github.com/jbosstools/jbosstools-fuse/pull/1424

bfitzpat commented:
Fixed minor issue with toolbar enablement. Still see a small issue where the new Rest Operation is selected, but the toolbar buttons are not being enabled properly (delete doesn't become active until you re-select the new operation). Will investigate further.

bfitzpat commented:
@lhein I'm wondering if you can take a look to see why the model changes are not being persisted back to the file on the back end. I can add the rest configuration, rest element, and an operation now, click Save, close the editor, and re-open it only to find the bits gone. That's the bigger issue than the toolbar enablement on the new rest operation creation.

lhein commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
false positive

lhein commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -688,10 +761,124 @@ public void handleEvent(Event event) {
 			selectedControl = newControl;
 			selection = getDataFromSelectedUIElement(newControl);
+			addRestOperationAction.setEnabled(!restList.getStructuredSelection().isEmpty());
+			deleteRestOperationAction.setEnabled(selection instanceof RestVerbElement);
```
false positive

lhein commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -547,6 +613,13 @@ private void refreshRestSection() {
+		
+		addRestOperationAction.setEnabled(
+				!ctx.getRestElements().isEmpty() && !restList.getStructuredSelection().isEmpty());
+		deleteRestOperationAction.setEnabled(selection instanceof RestVerbElement);
```
false positive

lhein commented:
fixed both the save issue and the button enablement issue.

apupier commented:
test in erro timed out on: 17:16:58 Running org.fusesource.ide.camel.editor.restconfiguration.RestEditorTestIT

lhein commented:
testPR

lhein commented:
applied the same fix to the properties page for rest operations that in #1424 for the rest elements

lhein commented:
rebased with master

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/internal/UIMessages.java:
```diff
@@ -24,6 +24,16 @@
+
+	public static String addRestOperationActionTooltip;
+	public static String addRESTOperationWizardErrorMessage;
+
```
this empty line seems useless

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/internal/UIMessages.java:
```diff
@@ -24,6 +24,16 @@
+	public static String addRESTOperationWizardErrorMessage;
+
+	public static String addRESTOperationWizardErrorTitle;
+
```
this empty line seems useless

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/internal/UIMessages.java:
```diff
@@ -202,6 +216,20 @@
+	public static String restVerbDefinitionPageIDCannotBeNullError;
+
+	public static String restVerbDefinitionPageIDField;
+
```
this empty line seems useless

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/internal/UIMessages.java:
```diff
@@ -202,6 +216,20 @@
+	public static String restVerbDefinitionPageIDField;
+
+	public static String restVerbDefinitionPageMessage;
+
```
this empty line seems useless.... and some others

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/properties/rest/RestVerbElementPropertiesSection.java:
```diff
@@ -85,15 +82,21 @@ private AbstractParameterPropertyUICreator handleField(final Parameter p, final
-			targetValueMap.put(p.getName(), targetValue);
+			targetValue.addChangeListener(event -> refreshClassAndBeanRefBindings());
+			// read only for now
+			creator.getControl().setEnabled(false);
```
I think there was a problem with the rebase, this code was removed to have properties editable for rest operations

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -325,33 +299,46 @@ private void getImages() {
+	private ToolBar createRestOperationTabToolbar(Composite parent) {
+		ToolBarManager toolBarManager = new ToolBarManager(SWT.FLAT);
+		ToolBar toolbar = toolBarManager.createControl(parent);
+		toolbar.setBackground(Display.getCurrent().getSystemColor(SWT.COLOR_WHITE));
```
need to use FormToolkit to create/adapt toolbar to have it coherent with possible different Eclipse theme

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
what do you think to use:
`List<AbstractCamelModelElement> toDelete = new ArrayList<>(ctx.getRestElements().values());`
instead of
```
ArrayList<AbstractCamelModelElement> toDelete = new ArrayList<>();
Iterator<AbstractCamelModelElement> iter = ctx.getRestElements().values().iterator();
while (iter.hasNext()) {
	toDelete.add(iter.next());
}
```

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/actions/AddRestElementAction.java:
```diff
@@ -0,0 +1,46 @@
+	
+	/**
+	 * 
+	 * @param imageReg
```
javadoc comment is inaccurate and provide no information
either remove or fix and provide more information

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/actions/AddRestOperationAction.java:
```diff
@@ -0,0 +1,46 @@
+	
+	/**
+	 * 
+	 * @param imageReg
```
javadoc comment is inaccurate and provide no information
either remove or fix and provide more information

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/actions/DeleteRestElementAction.java:
```diff
@@ -0,0 +1,46 @@
+	
+	/**
+	 * 
+	 * @param imageReg
```
javadoc comment is inaccurate and provide no information
either remove or fix and provide more information

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/actions/DeleteRestOperationAction.java:
```diff
@@ -0,0 +1,46 @@
+	
+	/**
+	 * 
+	 * @param imageReg
```
javadoc comment is inaccurate and provide no information
either remove or fix and provide more information

bfitzpat commented:
Addressed a number of issues, including some problems with setting the inner "to" element within a REST operation.

apupier commented:
2 tests related to REST are failing:
![image](https://user-images.githubusercontent.com/1105127/43078241-308f2720-8e8a-11e8-90c2-a85bbb850ece.png)

```
java.lang.NullPointerException
	at org.fusesource.ide.camel.editor.restconfiguration.RestEditorTestIT.openCamelFileRemoveRestOperationTest(RestEditorTestIT.java:317)
```

```
Expecting:
 <true>
to be equal to:
 <false>
but was not.
	at org.fusesource.ide.camel.editor.restconfiguration.RestEditorTestIT.openCamelFileMoveToRESTPage(RestEditorTestIT.java:80)
```

bfitzpat commented:
https://github.com/jbosstools/jbosstools-fuse/pull/1424 must be merged before this, but looking good!

bfitzpat commented:
* rebased PR after merging https://github.com/jbosstools/jbosstools-fuse/pull/1424
* Added https://issues.jboss.org/browse/FUSETOOLS-2898 to Epic to handle in later sprint
* Created https://issues.jboss.org/browse/FUSETOOLS-3051 to handle validating operation IDs in add operation wizard
* Updated https://issues.jboss.org/browse/FUSETOOLS-3049 to look into selection issue with REST element
* also fixed a number of other issues - https://github.com/jbosstools/jbosstools-fuse/pull/1426/commits/309dfa03e1a049c16da93a72486c5b2c595cbf45

apupier commented:
I think that the issue with the already existing ID must be handled in this PR, otherwise the user can have a popup with an error which is not comprehensible.
![image](https://user-images.githubusercontent.com/1105127/43185265-b01a5f5c-8feb-11e8-8be5-45c293ef2e4b.png)

bfitzpat commented:
The selection issue is an interesting one. Not quite sure yet what's causing that weirdness.

bfitzpat commented:
Addressed remaining issues with selection (FUSETOOLS-3049) and ID validation (FUSETOOLS-3051)
